Create a CupertinoScaffold

xster opened the pull request:
#10466

In this PR:
- Basic layout structure, positions nav bar and tab bar in it
- A tabbed content area that connects with a tab bar automatically

Not in this PR:
- Navigation and gestures
- Different navigator stack per tab

xster committed:
works using pageview
it’s a bit heavy and scrolls
no documentation or test yet

xster committed:
docs and tests

xster committed:
more docs tweak

xster commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+            (int index) {
+              return new Builder(
+                builder: (BuildContext context) {
+                  return widget.rootPageBuilder(context, index);
```
This works but isn't very efficient. The IndexedStack builds all of them right at the beginning. Don't know if there's something reusable to not build them until needed.

Hixie replied:
Yeah we definitely don't want to use an IndexedStack here.
...
...
(long in-person discussion)
...
...
We concluded we want to use a Stack, lazily inflate the children so that you only inflate them once they've been picked, then keep them in the tree, and you want to use TickerMode and Offstage to keep them cheaper.

Hixie replied:
Maybe instead of a Stack use a CustomMultiChildLayout so you can stop them being laid out as well.

xster replied:
Maybe I'm misapplying it. Did you mean give it a MultiChildLayoutDelegate and have it just layout and position the active tab only? I'm getting 
```
The _TabViewLayout custom multichild layout delegate forgot to lay out the following child:
  1: RenderOffstage#1048774387 NEEDS-LAYOUT NEEDS-PAINT
Each child must be laid out exactly once.
```
Or did you just mean to layout all the children with the delegate but don't do any complex calculations, just let them have the size of the parent all the time?

xster committed:
revert drive by page view changes

Hixie commented on packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:
```diff
@@ -31,8 +31,8 @@ const Color _kDefaultTabBarBorderColor = const Color(0x4C000000);
 // TODO(xster): document using with a CupertinoScaffold.
-class CupertinoTabBar extends StatelessWidget {
   /// Creates a tab bar in the iOS style.
+class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
```
i think you got a merge error here

Hixie commented on packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:
```diff
@@ -175,4 +180,30 @@ class CupertinoTabBar extends StatelessWidget {
+
+  /// Clone the current [CupertinoTabBar] with the [currentIndex] overridden
+  /// by `withCurrentIndex` and the [onTap] chained after `withOnTap` if present.
+  CupertinoTabBar clone({
```
we normally call this copyWith()

Hixie replied:
...and list all the arguments. Search for examples of "copyWith" to crib from.

Hixie commented on packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:
```diff
@@ -175,4 +180,30 @@ class CupertinoTabBar extends StatelessWidget {
+       currentIndex: withCurrentIndex,
+       // Concatenate the 2 callback functions.
+       onTap: (int newIndex) {
+         withOnTap(newIndex);
```
this seems dangerous... what if you keep applying it to the object you get back? Eventually you'll have this really long chain of callbacks...

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+import 'bottom_tab_bar.dart';
+import 'nav_bar.dart';
+
+/// Implements the basic iOS application's layout and behavior structure.
```
s/the/a/

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+  /// are required. The [CupertinoScaffold] will automatically listen to the
+  /// provide [CupertinoTabBar]'s tap callbacks to change the active tab.
+  ///
+  /// Tabs' contents is built with the provided [rootPageBuilder] at the active
```
are built

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+       rootPageBuilder = null,
+       super(key: key);
+
+  /// Construct a [CupertinoScaffold] with tabs. A [tabBar] and a [rootPageBuilder]
```
Usually you want the first paragraph to be just one sentence. The first paragraph is what ends up being used as the "short doc" in the API doc summaries.

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+       child = null,
+       super(key: key);
+
+  /// The [navigationBar], typically a [CupertinoNavigationBar] is drawn at the
```
missing comma before "is"

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+  /// that lets the user switch between different tabs in the main content area
+  /// when present.
+  ///
+  /// When provided, [CupertinoTabBar.currentIndex] will be ignored and and will
```
and and

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+  /// be managed by the [CupertinoScaffold] to show the currently selected page
+  /// as the active item index. [CupertinoTabBar.onTap] will still be called but
+  /// chained after an automatic handler from [CupertinoScaffold] to change
+  /// the current tab in the main content area.
```
this last sentence is confusing. I'm not sure it's worth including, honestly, but if you do want to be explicit about this I think it needs to be clearer what this means.

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+  final CupertinoTabBar tabBar;
+
+  /// An [IndexedWidgetBuilder] that's called when tabs become active. Used
+  /// when a tabbed scaffold is constructed via the [.tabbed] constructor.
```
`[.tabbed]` should be `[new CupertinoScaffold.tabbed]` I believe

xster commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,157 @@
+      stacked.add(_padMiddle(widget.child));
+    } else if (widget.rootPageBuilder != null) {
+      stacked.add(_padMiddle(
+        new IndexedStack(
```
offline chat summary: 
Make a custom private mechanism that uses offstage, tickermode and a custom lazy builder

motivations:
- lazy build initially
- stop animations for offscreen
- keep offscreen subtrees alive

xster committed:
swap out the indexed stack for a custom thing

xster reviewed (commented):
Not really ready for review again yet. Needs more testing. Just sending a batch review replies (and pushing from my home repo)

xster committed:
layout everything on the cheap
tests pass

xster commented:
Ummm it doesn't properly update to my branch anymore once I close it. I'll reopen it, mainly to get some broader feedback. The classes documentations aren't updated yet.

xster committed:
add more tests

Hixie commented:
I guess I was wrong about MultiChildCustomLayout. Disregard me. :-)

xster committed:
move back to stack and add docs

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,212 @@
+    this.navigationBar,
+    @required this.child,
+  })
+      : tabBar = null,
```
nit: here and below, put the colon on the same line as the `})`, as in
```dart
}) : tabBar = null,
     rootTabPageBuilder = null,
...
```

xster replied:
Done. I must have triggered the formatter without selecting a text range at some point

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,212 @@
+    return new Stack(
+      children: new List<Widget>.generate(widget.tabNumber, (int index) {
+        final bool active = index == widget.currentTabIndex;
+        // If the tab is being shown re-build and cache it.
```
We should also rebuild it if we've ever built it before, otherwise we'll be leaving stale widgets in the tree and no way to remove them (e.g. suppose one is a video, and the build function only includes the video if the tab is active -- if you don't rebuild the tab, you'll be leaving an active video and no way to remove it).

xster replied:
I was hoping we can keep it as simple as possible let the user handle edge cases by listening to their own supplied tab bar's onTap callbacks if needed. 

But I think I may have to address it anyway with some version of solution to the problem you're describing in the next PR once I start putting in the Navigators. I think I'd like to tie that API together with the various page routes' maintainState concepts etc.

Adding todo to the issues

Hixie replied:
These aren't edge cases. The contract when you have a builder is that you'll call it any time you build. Putting in old widgets is going to be a weird quirk that will lead to bugs. We should just call the builders.

It's fine to not call them until you need them, but once you're including them, you should call them (or not use a builder).

xster replied:
ok, done.

Hixie commented on packages/flutter/test/cupertino/scaffold_test.dart:
```diff
@@ -0,0 +1,205 @@
+List<int> selectedTabs;
+
+void main() {
+
```
nit: extraneous space

xster replied:
done

Hixie commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,212 @@
+  ///
+  /// Content can slide under the [navigationBar] or the [tabBar] when they're
+  /// translucent.
+  final Widget child;
```
the docs for child and rootTabPageBuilder should clarify that they can't both be non-null, that one must be non-null, and they should each point to the relevant constructor.

xster replied:
done

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

xster commented:
For posterity, moved back to Stack for readability simplicity. Tweaked docs and added tests. 
Performance behaviour looks as expected. 
![ky2bgr2dwby](https://user-images.githubusercontent.com/156888/27051850-7e2e190e-4f6c-11e7-87f5-24c6e2597a0a.png)

No frames are triggered until tabs are changed and only one frame is redrawn per tab.

xster committed:
review notes

xster commented on packages/flutter/lib/src/cupertino/scaffold.dart:
```diff
@@ -0,0 +1,212 @@
+    double topPadding = MediaQuery
+        .of(context)
+        .padding
+        .top;
```
oops, drive by change. I gotta find a way to disable the formatter

xster committed:
build all the tabs once built

Hixie commented:
(This raised our technical debt by about $9,000. This is fine, just reporting this for transparency.)

xster commented:
Ha, was it all the TODOs? Hope the interest rates aren't too high for a week or so.

Hixie commented:
Yup. :-)
